app/src/main/java/org/ea/sqrl/activites/UrlLoginActivity.java

```diff
@@ -132,36 +130,19 @@ public void afterTextChanged(Editable s) {
             long currentId = SqrlApplication.getCurrentId(this.getApplication());
 
             if(currentId != 0) {
-                doLogin(storage, txtLoginPassword, false, true, null,this);
+                doLogin(storage, txtLoginPassword, false, useCps, true, null,this);
             }
         });
 
-        if(Build.VERSION.SDK_INT >= Build.VERSION_CODES.P && storage.hasBiometric()) {
+        if(Build.VERSION.SDK_INT >= Build.VERSION_CODES.P && storage.hasBiometric() && !ACTION_LOGON.equals(intent.getAction())) {
             BioAuthenticationCallback biometricCallback =
-                    new BioAuthenticationCallback(UrlLoginActivity.this.getApplicationContext(), () -> {
-                        handler.post(() -> showProgressPopup());
-                        communicationFlowHandler.addAction(CommunicationFlowHandler.Action.QUERY_WITHOUT_SUK);
-                        communicationFlowHandler.addAction(CommunicationFlowHandler.Action.LOGIN_CPS);
-
-                        communicationFlowHandler.setDoneAction(() -> {
-                            storage.clear();
-                            handler.post(() -> {
-                                hideProgressPopup();
-                                closeActivity();
-                            });
-                        });
-
-                        communicationFlowHandler.setErrorAction(() -> {
-                            storage.clear();
-                            handler.post(() -> hideProgressPopup());
-                        });
-
-                        communicationFlowHandler.handleNextAction();
-                    });
+                    new BioAuthenticationCallback(UrlLoginActivity.this.getApplicationContext(), () ->
+                        doLogin(storage, txtLoginPassword, false, useCps, false, UrlLoginActivity.this, UrlLoginActivity.this)
+                    );
 
             BiometricPrompt bioPrompt = new BiometricPrompt.Builder(this)
                     .setTitle(getString(R.string.login_title))
-                    .setSubtitle(domain)
+                    .setSubtitle(mSqrlMatcher.group(1))
```

Comment:
See comment above.

Reply:
Yes, I was thinking of adding a `getDomain()` method to the matcher to get around this, but I like the idea of restoring the `String` definition and moving all the biometric code.

Just a wild, but maybe related idea... you know those stack traces thrown because biometric code is inside an activity?  If you do go about moving the biometric stuff, you might consider looking at the _concept_ used in `LoginActivityHelper` (https://github.com/sengsational/secure-quick-reliable-login/tree/remove_starting_stacktraces) to prevent those stack traces.  That code was never merged, because it's not really "our fault" that these stack traces are thrown.  Still, if it makes sense, pulling the biometrics out of the activity itself might be an option.

Reply:
Good point, Dale. If I remember correctly, the PR was never merged because there was some other controversial stuff in it. The biometric stuff everyone was fine with as far as I remember. 

Me personally, I would appreciate pulling the biometrics stuff out into its own helper class, but would then call it something like `BiometricLoginHelper` to better reflect what it acutally does.

However, I think we should not be adding more stuff into this PR at this time, but instead do all of these things as small follow-up PRs that are not tangled up with the consolidation stuff which is at the center of this PR, and will therefore be much easier to review and test.

Reply:
Agreed.